### PR TITLE
feat: SSA based dependent resource matching and create/update

### PR DESCRIPTION
**Pull request by csviri**
The PR has two main parts:
1. changed the default way how the dependent resources create/update the target resource. From now on it will use by default server side apply. 

2. It provides a new default matcher the `SSABasedGenericKubernetesResourceMatcher`. The intention with this new matcher is to provide a better default. This won't be perfect, thus for sure there are corner cases that are not covered. Either we will patch it or just advise to implement a specific matcher for the users use-case.

### CODE REVIEW

**Comment by csviri** [comment]
@metacosm will introduce a separate feature flag for matching, and have that set for legacy while have set the update/create to SSA. I think that is a good compromise.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        keepOnlyManagedFields(emptyResMapValue, value, managedEntryByIndex.get(key), objectMapper);
+      } catch (JsonProcessingException ex) {
+        throw new IllegalStateException(ex);
+      }
+    });
+  }
+
+  // set values, the "v:" prefix
+  @SuppressWarnings("rawtypes")
+  private static void handleSetValues(Map<String, Object> result, Map<String, Object> actualMap,
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:212]: Please add documentation about what this method is doing and why.

**Review by metacosm — CHANGES_REQUESTED** [review]
LGTM as far as I can tell without diving into the managed fields structure. This is why it's important to document some of the steps so that future maintainers will be able to understand some of the things the current implementation does, in particular with respect to the more exotic fields.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    metadata.remove(NAMESPACE_KEY);
+    if (metadata.isEmpty()) {
+      desiredMap.remove(METADATA_KEY);
+    }
+    desiredMap.remove(KIND_KEY);
+    desiredMap.remove(API_VERSION_KEY);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void keepOnlyManagedFields(Map<String, Object> result,
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:118]: probably is, just would need a refactor. If you insist can do that. A separate PR?

**Comment by csviri** [comment]
> LGTM as far as I can tell without diving into the managed fields structure. This is why it's important to document some of the steps so that future maintainers will be able to understand some of the things the current implementation does, in particular with respect to the more exotic fields.

yep, tbh there are some corner cases that are probably not covered, but should be very rare, I expect that we will need to do some forward fixes. But will document the the current one more in depth according to your comments.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java (hunk 1 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:225]: added docs

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+  private static boolean isSetValueField(Set<Map.Entry<String, Object>> managedEntrySet) {
+    return isKeyPrefixedSkippingDotKey(managedEntrySet, V_PREFIX);
+  }
+
+  private static boolean isListKeyEntrySet(Set<Map.Entry<String, Object>> managedEntrySet) {
+    return isKeyPrefixedSkippingDotKey(managedEntrySet, K_PREFIX);
+  }
+
+  private static boolean isKeyPrefixedSkippingDotKey(Set<Map.Entry<String, Object>> managedEntrySet,
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:273]: added docs

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            f -> f.getManager().equals(fieldManager) && f.getOperation().equals(APPLY_OPERATION))
+        .collect(Collectors.toList());
+    if (targetManagedFields.isEmpty()) {
+      log.debug("No field manager exists for resource {} with name: {} and operation Apply ",
+          actual, actual.getMetadata().getName());
+      return Optional.empty();
+    }
+    // this should not happen in theory
+    if (targetManagedFields.size() > 1) {
+      log.debug("More than one field manager exists with name: {} in resource: {} with name: {} ",
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:344]: changed

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        (Map<String, Object>) managedFields.get(key), objectMapper);
+  }
+
+  private static boolean isNestedValue(Map<?, ?> managedFieldValue) {
+    return !managedFieldValue.isEmpty();
+  }
+
+  // list entries referenced by key, or when "k:" prefix is used
+  @SuppressWarnings("unchecked")
+  private static void handleListKeyEntrySet(Map<String, Object> result,
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/SSABasedGenericKubernetesResourceMatcher.java:187]: added

**Review by metacosm — APPROVED** [review]

**Pull request merged**